distributed/shuffle/_shuffle_extension.py

```diff
@@ -194,37 +204,40 @@ async def receive(self, data: list[bytes]) -> None:
         await self._receive(data)
 
     async def _receive(self, data: list[bytes]) -> None:
-        if self._exception:
-            raise self._exception
+        self.raise_if_closed()
 
         try:
             self.total_recvd += sum(map(len, data))
-            # TODO: Is it actually a good idea to dispatch multiple times instead of
-            # only once?
-            # An ugly way of turning these batches back into an arrow table
-            data = await self.offload(
-                list_of_buffers_to_table,
-                data,
-                self.schema,
-            )
+            groups = await self.offload(self._repartition_buffers, data)
+            await self._write_to_disk(groups)
+        except Exception as e:
+            self._exception = e
+            raise
 
-            groups = await self.offload(split_by_partition, data, self.column)
+    def _repartition_buffers(self, data: list[bytes]) -> dict[str, list[bytes]]:
+        table = list_of_buffers_to_table(data, self.schema)
+        groups = split_by_partition(table, self.column)
+        assert len(table) == sum(map(len, groups.values()))
+        del data
+        return {
+            k: [batch.serialize() for batch in v.to_batches()]
+            for k, v in groups.items()
+        }
 
-            assert len(data) == sum(map(len, groups.values()))
-            del data
+    async def _write_to_disk(self, data: dict[str, list[bytes]]) -> None:
+        self.raise_if_closed()
+        await self._disk_buffer.write(data)
 
-            groups = await self.offload(
-                lambda: {
-                    k: [batch.serialize() for batch in v.to_batches()]
-                    for k, v in groups.items()
-                }
+    def raise_if_closed(self) -> None:
+        if self.closed:
+            if self._exception:
+                raise self._exception
+            raise ShuffleClosedError(
+                f"Shuffle {self.id} has been closed on {self.local_address}"
             )
```

Comment:
Not raising here at all will currently only fail two tests. This feels a bit thin considering that this is invoked 13 times.

Reply:
`test_closed_input_only_worker_during_transfer`
`test_crashed_input_only_worker_during_transfer` 
are failing. I believe this is only a different exception.

`test_closed_worker_between_repeats` failed with a deadlock/timeout

Reply:
sometimes all of them are green